File: 236. Lowest Common Ancestor of a Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        '''
        Solution easy to follow
        '''
        self.ans = None
        self.trav_tree(root, p, q)
        return self.ans
        
        
    def trav_tree(self, root, p, q):
        if not root:
            return False
        left = self.trav_tree(root.left, p, q)
        right = self.trav_tree(root.right, p, q)
        mid = root == p or root == q
        if left+right+mid >=2:
            self.ans = root
        return left or right or mid

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        '''
        more concise one
        '''
        if not root or root in [p, q]:
            return root
        left = self.lowestCommonAncestor(root.left, p, q)
        right = self.lowestCommonAncestor(root.right, p, q)
        if not left:
            return right
        elif not right:
            return left
        else:
            return root

File: 236. Lowest Common Ancestor of a Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
/**
 * dfs without saving nodes, time O(n) space O(h)
 */
class Solution {
    TreeNode ans;
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        tt(root, p, q);
        return ans;
    }
    private boolean tt(TreeNode node, TreeNode p, TreeNode q) {
        if (node == null) {
            return false;
        }
        int left = tt(node.left, p, q) ? 1 : 0;
        int right = tt(node.right, p, q) ? 1 : 0;
        int mid = (node==p || node==q) ? 1 : 0;
        if (left+right+mid >= 2) {
            ans = node;
        }
        return (left+right+mid) >= 1;
    }
}
/**
 * dfs with saving path, time O(n) space O(height)
 */
class Solution {
    List<TreeNode> lp, lq;
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        dfs(root, p, q, new ArrayList<>());
        for (int i=0;i<Math.min(lp.size(), lq.size());i++) {
            if (lp.get(i) != lq.get(i)) {
                return lp.get(i-1);
            }
        }
        return lp.get(Math.min(lp.size(), lq.size())-1);
    }
    private void dfs(TreeNode node, TreeNode p, TreeNode q, List<TreeNode> temp) {
        if (node == null) {
            return;
        }
        temp.add(node);
        if (node == p) {
            lp = new ArrayList(temp);
        }
        if (node == q) {
            lq = new ArrayList(temp);
        }
        dfs(node.left, p, q, temp);
        dfs(node.right, p, q, temp);
        temp.remove(temp.size()-1);
    }
}